feat: implement mismatch record sampling

shsk opened the pull request:
Add functionality to sample mismatch records using random or stratified sampling methods16. This feature enhances the reconcile project by providing a representative subset of mismatches for analysis.

sundarshankar89 commented on src/databricks/labs/remorph/reconcile/sampler.py:
```diff
@@ -0,0 +1,199 @@
+            raise ValueError("RandomSampler: Only 'random' sampling method is supported")
+
+        specs = self._sampling_options.specifications
+        if specs.type == SamplingSpecificationsType.FRACTION and (specs.value is None or (not 0 < specs.value < 1)):
```
Move this config validation into post_init of `SamplingOptions` class so we fail the process when the job is initiated than after we fetch

shsk replied:
Change Implemented in the latest commit.

sundarshankar89 commented on src/databricks/labs/remorph/reconcile/sampler.py:
```diff
@@ -0,0 +1,199 @@
+        return sampled_df.select(*key_columns)
+
+
+class SamplerFactory:
```
Add a TODO here. I would like to move away from the factory to something like 'ReconcileContext` to replace all factory implementations. this can be Tech Debt for future.

shsk replied:
Change Implemented in the latest commit.

sundarshankar89 commented on src/databricks/labs/remorph/reconcile/sampler.py:
```diff
@@ -0,0 +1,199 @@
+        return sampled_df
+
+
+class StratifiedSampler(Sampler):
```
nit: Can you add Doc String on what `StratifiedSampler` means?
in the subsequent PR we can make changes to Docs

shsk replied:
Change Implemented in latest commit.

sundarshankar89 commented on tests/integration/reconcile/test_sampler.py:
```diff
@@ -0,0 +1,233 @@
+    )
+    random_sampling_options = SamplingOptions(
+        method= SamplingOptionMethod.RANDOM,
+        specifications= SamplingSpecifications(type=SamplingSpecificationsType.COUNT, value=2),
```
nit: again 

```suggestion
        specifications= SamplingSpecifications(type=SamplingSpecificationsType.COUNT, value=2.0),
```

sundarshankar89 replied:
When we enable stricter Lint Checks this call will fail.

shsk replied:
Fixed in the latest commit.

sundarshankar89 commented on src/databricks/labs/remorph/reconcile/sampler.py:
```diff
@@ -0,0 +1,199 @@
+        if sampling_options is None:
+            default_sampling_options = SamplingOptions(
+                method=SamplingOptionMethod.RANDOM,
+                specifications=SamplingSpecifications(type=SamplingSpecificationsType.COUNT, value=100),
```
```suggestion
                specifications=SamplingSpecifications(type=SamplingSpecificationsType.COUNT, value=_MIN_SAMPLE_COUNT),
```

shsk replied:
Change Implemented in latest commit.

sundarshankar89 commented on tests/integration/reconcile/test_sampler.py:
```diff
@@ -0,0 +1,233 @@
+    # Perform random sampling
+    random_sample = random_sampler.sample(keys_df, ["key"], target_table_df)
+
+    assert random_sample.count() > 0
```
```suggestion
    assert random_sample.count() == 2
```

shsk replied:
Fixed in the latest commit.

sundarshankar89 commented on tests/integration/reconcile/test_sampler.py:
```diff
@@ -0,0 +1,233 @@
+    # Perform random sampling
+    random_sample = random_sampler.sample(keys_df, ["key"], target_table_df)
+
+    assert random_sample.count() > 0
```
Same as above

shsk replied:
Fixed in the latest commit.

sundarshankar89 reviewed (approved):
Few nit, good frist draft

shsk commented:
Added functionality to sample mismatch records using random or stratified sampling methods. This feature enhances the reconciliation project by providing a representative subset of mismatches for analysis.

Fixes:
1. Fixed Formatting issues
2. Fixed integration test cases
3. Incorporated changes for review comments
4. Corrected typos in comments

shsk commented:
Fixes:
1. Fixed Formatting issues
2. Fixed integration test cases
3. Incorporated changes for review comments
4. Corrected typos in comment

sundarshankar89 reviewed (approved):
LGTM

sundarshankar89 commented:
@bishwajit-db Can you stamp it.

bishwajit-db reviewed (approved):
LGTM

shsk commented:
Found and issue related to column mapping with Stratified sampling, hence converting to draft.

gueniai commented:
Hi @shsk , we require that all commits are verified. In order to do that, unfortunately you'll need to re-do this PR after you go through this: https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification 
It's a one time setup and after that all your labs contributions will be good to go. Thank you!
